crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs

```diff
@@ -156,69 +160,127 @@ pub(super) fn configure_dsn(
         networking_parameters_registry,
         request_response_protocols: vec![
-            PieceByHashRequestHandler::create(move |&PieceByHashRequest { piece_index_hash }| {
-                debug!(?piece_index_hash, "Piece request received. Trying cache...");
-                let multihash = piece_index_hash.to_multihash();
+            PieceAnnouncementRequestHandler::create({
+                let provider_record_announcer = provider_record_announcer.clone();
+
+                move |peer_id, req| {
+                    trace!(?req, %peer_id, "Piece announcement request received.");
+
+                    let mut provider_storage = farmer_provider_storage.clone();
+                    let provider_record_announcer = provider_record_announcer.clone();
+                    let req = req.clone();
+
+                    async move {
+                        let key = match req.piece_key.clone().try_into() {
+                            Ok(key) => key,
+
+                            Err(error) => {
+                                error!(
+                                    %error,
+                                    %peer_id,
+                                    ?req,
+                                    "Failed to convert received key to record:Key."
+                                );
 
-                let weak_readers_and_pieces = weak_readers_and_pieces.clone();
-                let piece_store = piece_store.clone();
-                let piece_memory_cache = piece_memory_cache.clone();
+                                return None;
+                            }
+                        };
 
-                async move {
-                    if let Some(piece) = piece_memory_cache.get_piece(&piece_index_hash) {
-                        return Some(PieceByHashResponse { piece: Some(piece) });
+                        let provider_record = ProviderRecord {
+                            provider: peer_id,
+                            key,
+                            addresses: req.converted_addresses(),
+                            expires: KADEMLIA_PROVIDER_TTL_IN_SECS.map(|ttl| Instant::now() + ttl),
+                        };
+
+                        if let Err(error) = provider_storage.add_provider(provider_record.clone()) {
+                            error!(
+                                %error,
+                                %peer_id,
+                                ?req,
+                                "Failed to add provider for received key."
+                            );
+
+                            return None;
+                        }
+
+                        if let Some(provider_record_announcer) =
+                            provider_record_announcer.read().as_ref()
+                        {
+                            provider_record_announcer.announce(&provider_record);
+                        }
+
+                        Some(PieceAnnouncementResponse)
```

Comment:
I would have expected this to be the same as `node.on_announcement`, why is it completely different?